### ISSUE TITLE
Weird range behavior

### ISSUE THREAD

**Issue by KAAtheWiseGit** [nushell/nushell#11086]
### Describe the bug

Writing two consecutive ranges causes weird output.

It seems to be an undocumented feature, I think?

### How to reproduce

- `1..2..100` enumerates 1-100.
- `1..3..100` enumerates 1-99 with step 2.
- `1..(-1)..100` gives a "can't convert to countable values" error, which I cannot decipher.

### Expected behavior

I had no idea ranges could do that.  I think this functionality should either be documented or removed.  This also resembles Haskell ranges, which I personally find less obvious than Python `start:end:step` ones.

### Configuration

| key                | value                                                                                                               |
| ------------------ | ------------------------------------------------------------------------------------------------------------------- |
| version            | 0.86.0                                                                                                              |
| branch             |                                                                                                                     |
| commit_hash        |                                                                                                                     |
| build_os           | linux-x86_64                                                                                                        |
| build_target       | x86_64-alpine-linux-musl                                                                                            |
| rust_version       | rustc 1.72.1 (d5c2e9c34 2023-09-13) (Alpine Linux 1.72.1-r0)                                                        |
| cargo_version      | cargo 1.72.1                                                                                                        |
| build_time         | 2023-10-22 21:58:37 +00:00                                                                                          |
| build_rust_channel | release                                                                                                             |
| allocator          | mimalloc                                                                                                            |
| features           | default, sqlite, trash, which, zip                                                                                  |
| installed_plugins  | custom-value generate, custom-value generate2, custom-value update, gstat, nu-example-1, nu-example-2, nu-example-3 |

**Comment by fdncred** [comment]
yup, i've always thought the 1..3..100 functionality was bad. I prefer seq's seq 1 3 100, which is "count by 3's". it's much less obvious to me. it's like some functional pl nerd flexing.

1..3..100 means 3 minus 1 is 2 so count by 2's starting at 1.
1..6..100 means 6 minus 1 is 5 so count by 5's starting at 1.

I prefer
seq 1 3 100 which means count from 1 to 100 by 3's.

**Comment by KAAtheWiseGit** [comment]
@fdncred I found it weird it was undocumented.  Honestly, I don't get the utility of ranges in Nushell in general, since they can't be used for their primary use in other programming languages, slicing lists (`[1 2 3 4 5] | take/get 1..3`).

**Comment by fdncred** [comment]
I mostly use it like in the range command for filtering outliers.
```nushell
$data | sort | range 2..(-3)
```

So, assuming `$data` is like a list of sorted durations, the pipeline above will remove the first 2 and last 2 ouliers.

I also use it when I need iteration sometimes like `0..5 | each { blah }`. I'm sure there are some nu_scripts that I've written that use it.

**Comment by KAAtheWiseGit** [comment]
Yeah, I forgot about the iteration, including `for`.  I'd prefer a `range` function, but its just my minor bike shedding.

I think adding the `x..y..z` to the docs is the way to go.

**Comment by IanManske** [comment]
Yeah, when I discovered the `x..y..z` range behavior I was confused as well. Documenting it as you said would be nice.
While I can see some cases where ranges are useful, I don't like the fact that ranges may have a increment. Like what does `random int 0..2..10`  or `range 0..2..10` mean? I think none of the internal commands account for the case when a range has a step.

**Comment by amtoine** [comment]
yeah +1 that the following is confusing :open_mouth: 
```nushell
> 1..100 == 1..2..100
true
```

**Comment by sophiajt** [comment]
I think we did it for ease of parsing, using `1,2..10` would read a bit better, but might be a bit harder to parse?

**Comment by amtoine** [comment]
@sophiajt 
if i understood correctly, what is really confusing is that i expected `1..2..100` to list from `1` to `100` by steps of `2` but the step was `1`, i.e. it did list all of the integers from `1` to `100` :confused:

**Comment by KAAtheWiseGit** [comment]
Languages with `start..end` and without separate syntax for stepping:

- [Rust](https://doc.rust-lang.org/reference/expressions/range-expr.html).
- [Zig](https://ziglang.org/documentation/master/#Slices).
- [Ruby](https://www.rubyguides.com/2016/06/ruby-ranges-how-do-they-work/), it has `(10..20).step(2)`.
- [Crystal](https://crystal-lang.org/reference/1.10/syntax_and_semantics/literals/range.html).
- [Fish shell](https://fishshell.com/docs/current/language.html#slices).  They also allow combining slices with `$arr[1..3 10..13]`.

`start..end` with an increment at the end:

- [Python](https://docs.python.org/3.8/reference/expressions.html#slicings)
- [Bash](https://linuxize.com/post/bash-sequence-expression/).
- [Kotlin](https://kotlinlang.org/docs/ranges.html#progression).  Sort of.  It has `1..8 step 2`, where `step` is a function on the range object.
- [Elixir](https://hexdocs.pm/elixir/1.12/Range.html) with a somewhat quirky `start..end//step`.

A separate range object:

- [D](https://dlang.org/library/std/range.html).
- [Scala](https://docs.scala-lang.org/overviews/collections-2.13/concrete-immutable-collection-classes.html#ranges).

"Arithmetic sequences".

- [Haskell](https://www.haskell.org/tutorial/goodies.html) (almost at the end).  The closest one to Nushell I found, with `[1,3..10] => [1,3,5,7,9]`.

**Comment by fdncred** [comment]
> [F](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/loops-for-in-expression). This is an interesting one. Here, it is start .. skip .. finish, so 1..2..10 will give 1 3 5 7 9.

This is how the *nix `seq` command works, which I ported from uutils many moons ago.

```nushell
❯ seq 1 2 10
╭───┬───╮
│ 0 │ 1 │
│ 1 │ 3 │
│ 2 │ 5 │
│ 3 │ 7 │
│ 4 │ 9 │
╰───┴───╯
```

Nushell also used to be able to combine ranges because I had scripts like `for x in 0..10 20..30 { blah }` but somewhere along the way that was removed.

I could live with start..end..step, i think it's more intuitive than what we currently have.

**Comment by dedebenui** [comment]
This is very niche, but the Grapher app in macOS, like Haskell, also uses the common math notation:

```
n = {1, 3, ..., 15}
# equivalent to {1, 3, 5, 7, 9, 12, 15}
```

I think this is very intuitive and a nice complement to the `seq` command.

**Comment by CAESIUS-TIM** [comment]
## Differences between `range` and `seq`

```nu
# [NOTE] `str join`
1..3 | str join '|'
# Error: command doesn't support range input
seq 1 3 | str join '|'
# 1|2|3
seq 1 3 | describe
# list<int> (stream)
1..3 | each {} | str join '|'
# 1|2|3
1..3 | each {} | describe
# list<int> (stream)

# [NOTE] Also, `skip` and so many
```

## Range is fast and can be used for range checks

```nu
let start = 1
let end = 1_000_000
let step = 3
let range = 1..1_000
timeit {
  for i in $range {
    let target = random int (-2_000_000..2_000_000)
    $target in $start..($start + $step)..$end
  }
} | print
# 1ms 254µs 786ns
timeit {
  for i in $range {
    let target = random int (-2_000_000..2_000_000)
    $target in (seq $start $step $end)
  }
} | print
# [NOTE] Of course it's slow — it's not lazy-loaded
# 7sec 417ms 831µs 372ns
timeit {
  for i in $range {
    let target = random int (-2_000_000..2_000_000)
    (
      $start <= $target and $target <= $end
      and (($target - $start) mod $step) == 0
    )
  }
} | print
# 1ms 545µs
```

## No empty range

```nu
1..-1..2
# Error: nu::shell::range_to_countable
#
#   × Can't convert range to countable values
#    ╭─[entry #173:1:1]
#  1 │ 1..-1..2
#    · ────┬───
#    ·     ╰── can't convert to countable values
#    ╰────
```
